### ISSUE TITLE
PVR2: programs erroneously attempting to re-open closed lists.

### ISSUE THREAD

**Issue by snickerbockers** [washingtondc-emu/washingtondc#9]
When the following patch applied to WashingtonDC, the abort statement will eventually trigger while running the png example in KallistiOS' examples directory (the one that displays scrolling text of President Lincoln's Gettysburg Address rendered via textured triangle strips on PVR2).

I don't believe this should be possible based on my reading of the example program; if this is true then there is a bug in WashingtonDC.  KallistiOS' pvr_wait_ready function will block until the interrupt is issued (or up to 100 ms if it never gets the signal) so it shouldn't be possible for it to re-open display lists it's already closed.

One possible root-cause may be that we're sending the render-complete interrupt immediately (see pvr2_ta_startrender in src/hw/pvr2/pvr2_ta.c); on a real dreamcast this would definitely take a quantifiable amount of time.

```
diff --git a/src/hw/pvr2/pvr2_ta.c b/src/hw/pvr2/pvr2_ta.c
index bfb1f2a..aafb36c 100644
--- a/src/hw/pvr2/pvr2_ta.c
+++ b/src/hw/pvr2/pvr2_ta.c
@@ -403,6 +403,7 @@ static void on_end_of_list_received(void) {
     } else {
         printf("Unable to close the current display list because no display "
                "list has been opened\n");
+        abort(); // DO NOT COMMIT
         return;
     }
 
```

**Comment by snickerbockers** [comment]
With the following patch applied:
```
diff --git a/src/hw/sh4/sh4_excp.c b/src/hw/sh4/sh4_excp.c
index b6948da..90f99c3 100644
--- a/src/hw/sh4/sh4_excp.c
+++ b/src/hw/sh4/sh4_excp.c
@@ -20,6 +20,8 @@
  *
  ******************************************************************************/
 
+#include <stdlib.h>
+
 #include "sh4.h"
 #include "sh4_excp.h"
 #include "error.h"
@@ -96,8 +98,9 @@ static Sh4ExcpMeta const sh4_excp_meta[SH4_EXCP_COUNT] = {
 void sh4_enter_exception(Sh4 *sh4, enum Sh4ExceptionCode vector) {
     struct Sh4ExcpMeta const *meta = NULL;
     reg32_t *reg = sh4->reg;
+    unsigned idx;
 
-    for (unsigned idx = 0; idx < SH4_EXCP_COUNT; idx++) {
+    for (idx = 0; idx < SH4_EXCP_COUNT; idx++) {
         if (sh4_excp_meta[idx].code == vector) {
             meta = sh4_excp_meta + idx;
             break;
@@ -109,6 +112,10 @@ void sh4_enter_exception(Sh4 *sh4, enum Sh4ExceptionCode vector) {
         RAISE_ERROR(ERROR_UNKNOWN_EXCP_CODE);
     }
 
+    printf("SH4 exception index %u\n", idx);
+    if (idx >= SH4_EXCP_COUNT)
+        abort();
+
     reg[SH4_REG_SPC] = reg[SH4_REG_PC];
     reg[SH4_REG_SSR] = reg[SH4_REG_SR];
     reg[SH4_REG_SGR] = reg[SH4_REG_R15];
```

this gets printed:
```
POLY HEADER PACKET!
END-OF-LIST PACKET!
Display list "Opaque" closed
opaque list complete
Opening display list Transparent
texture enabled
the texture format is 2
The texture address ix 0x00396720
not twiddled
texture dimensions are (256, 256)
Texture 0x00396720 found in cache
POLY HEADER PACKET!
frame_stamp 4043 rendered
render_next_geo_buf - 1 geo_bufs rendered
vblank-out interrupt
SH4 exception index 37
clearing ISTNRM bits 0x00000094
ISTNRM went from 0x00000094 to 0x00000000
that inclued the STARTRENDER bit
Reading 0 from ISTEXT
SH4 exception index 39
tex_enable should be true
POLY HEADER PACKET!
END-OF-LIST PACKET!
Display list "Transparent" closed
transparent list complete
SH4 exception index 41
vblank-in interrupt
vcount is 525
passing framebuffer dimensions=(640, 480)
interlacing is enabled
SH4 exception index 39
WARNING: unable to open list Opaque because it is already closed
next_poly_group - no lists are open
POLY HEADER PACKET!
ERROR: unable to render vertex because no display lists are open
ERROR: unable to render vertex because no display lists are open
ERROR: unable to render vertex because no display lists are open
ERROR: unable to render vertex because no display lists are open
WARNING: unable to open list Opaque because it is already closed
next_poly_group - no lists are open
POLY HEADER PACKET!
END-OF-LIST PACKET!
Unable to close the current display list because no display list has been opened
aborted (core dumped)
~/programs/washingtondc/build $ 
```

Exception index 39 would be SH4_EXCP_TMU0_TUNI0.  This is unexpected.

**Comment by snickerbockers** [comment]
> Exception index 39 would be SH4_EXCP_TMU0_TUNI0. This is unexpected.

One possible explanation would be that pvr_wait_ready timed out waiting on pvr_state.ready_sem via sem_wait_timed.  This would mean that the root-cause is that KallistiOS missed a render-finished interrupt.  That could be the result of a bug in my code, or it could be that the program missed the interrupt since it wasn't designed for a system where rendering completes instantly.

**Comment by snickerbockers** [comment]
This was fixed by commit afe9cf4103e6be3e2f7109d19454973614e2bbb3 .
